src/lib/queue.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * Super simple queue, that allows you to bind a method to a queue and it will run them in order.
+ */
+export class Queue {
+	private queue: {
+		isRunning: boolean
+		fcn: () => Promise<void>
+		resolve: () => void
+		reject: (err: Error) => void
+	}[] = []
+
+	public bindMethod<T extends (...args: any) => Promise<void>>(
+		fcn: T,
+		clearWaiting?: {
+			reason: string
+		}
+	): T {
+		return (async (...args: any[]) => {
+			await new Promise<void>((resolve, reject) => {
+				if (clearWaiting) {
+					this.clearWaiting(clearWaiting.reason)
+				}
+
+				this.queue.push({
+					isRunning: false,
+					fcn: async () => {
+						return fcn(...args)
+					},
+					resolve,
+					reject,
+				})
+
+				this.checkQueue()
+			})
+		}) as T
+	}
+
+	/**
+	 * Clear the queue from any not-yet-running functions and rejected them with the given reason.
+	 * Any functions that are already running will not be affected.
+	 */
+	public clearWaiting(reason: string): void {
```

Comment:
This should probably this be doing something like a `this.queue = []` after rejecting all of the promises